### ISSUE TITLE
Clarify what suppressing rendering means in terms of lifecycle updates

### ISSUE THREAD

**Issue by sebmarkbage** [WICG/view-transitions#171]
In Chrome there's a [heuristic to delay paint](https://web.dev/preload-optional-fonts/) for 100ms on initial page load if a font is preloaded and optional. This is great because it avoids flashes of invisible text.

I'm not sure how/if this part will ever be spec:ed. Ideally this could also be used with other display modes to avoid the CLS if possible. This is also related to the [font-display: critical proposal](https://github.com/w3c/csswg-drafts/issues/7271).

One problem with these heuristics is that they only work for initial load. For SPA transitions it doesn't work. If you navigate to a new page that uses a font there's no way to delay the paint because everything paints as soon as possible - when there's some pending DOM mutations.

We can do some stuff to delay the DOM mutations in frameworks but because detecting if you actually use something requires apply styles which require DOM mutations makes this tricky to implement efficiently and easy to use, because there's no way to create an optimistic DOM tree for what might be rendered unless you have a way to revert any changes.

However, shared-element-transitions do have some of those capabilities at the painting layer. You could effectively delay the animation from starting if painting the new element would be missing a font.

So my proposal is to allow shared element transitions to automatically delay playing the animation for up to 100ms if a new element entering or element with updated text styles/content is still waiting on a font-face to load - under the same heuristics that apply for initial paint. In other words, wait to paint the "target screenshot" until the fonts have loaded or some timeout.

Delaying animations might be somewhat controversial since it can delay the feeling of responsiveness but it can also be better sometimes. Therefore this could be some extra option to the transition that opts into waiting for some heuristic.

**Comment by jakearchibald** [comment]
Here's a demo:

No waiting for fonts: https://simple-set-demos.glitch.me/waiting-for-fonts/without-waiting/
With waiting for fonts: https://simple-set-demos.glitch.me/waiting-for-fonts/with-waiting/

The difference between the two is `await document.fonts.ready`.

**Comment by jakearchibald** [comment]
For others reading, this should be paired with a timeout of sorts, such as 100ms as @sebmarkbage suggested:

```js
const wait = ms => new Promise(r => setTimeout(r, ms));

const transition = document.createDocumentTransition();
transition.start(async () => {
  updateTheDOM();
  await Promise.race([document.fonts.ready, wait(100)]);
});
```

**Comment by sebmarkbage** [comment]
Neat!

Unfortunately, this doesn't seem to work with `font-display: optional` in my testing (e.g. if you switch to display=swap in the Google Font urls).

I suspect that's an under specified part of the spec because it's very subtle whether a font has been used or not.

E.g. without transitions you can use this technique (in Chrome at least) to ensure that the font usage gets detected but then is not "shown" and so it doesn't trigger the permanent fallback state:

```js
  const p = document.fonts.ready;
  document.body.style.display = "none";
  await p;
  document.body.style.display = "";
```

I suspect something similar is necessary for transitions.

**Comment by jakearchibald** [comment]
Yeah, in the current implementation, while the paint is 'held', the only thing we're skipping is updating the pixels on the screen. We're going to change that so the render steps don't run at all (as if the display was 0hz).

For example, right now, things like animated gifs will be 'running' in the background. With the change, it will ensure that the gif starts along with the first frame of the transition.

**Comment by vmpstr** [comment]
Just as a note, we need to verify that the gifs that will be running are going to be in the background. Chromium's implementation drives gif animations from the compositor, so it's not immediately clear to me that skipping paint or skipping rendering altogether would have an effect on those.

**Comment by jakearchibald** [comment]
@vmpstr does that mean compositor-driven CSS animations won't pause either? That seems a little inconsistent

**Comment by vmpstr** [comment]
Yes, I believe that's the case

**Comment by khushalsagar** [comment]
There is an open [issue](https://tabatkins.github.io/specs/css-shared-element-transitions/#issue-edbf8829) in the spec for clarifying what suppressing rendering means. The last meeting where we discussed this the conclusion was that all animations should be paused otherwise you have this inconsistency of which animation runs depending on the UA's implementation of what's threaded (like GIFs in Chrome's case).

@jakearchibald @vmpstr does that conclusion align with you both?

**Comment by jakearchibald** [comment]
+1

**Comment by mmocny** [comment]
Late to this conversation, but, in what ways is this related to image decoding=sync?

As far as I know decoding=sync will also "block rendering", and also somewhere down stream from Paint (at least in chromium), but does not block declarative compositor-driven animations/scrolling updates.  I think that means that "Activate" stage is blocked?  And it may also have a timeout?

**Comment by vmpstr** [comment]
(All in Chromium terms) You're right that decoding=sync will block activation until the rasterization of affected images is done. This does mean that ongoing active tree animations will continue producing frames. 

For this feature, we're blocking both main and impl frames which means that even the active tree will not produce new frames, after the initial one that requests the snapshot which needs to be forwarded to the gpu stack. This means that active animations, gifs, etc will all be "paused"

**Comment by vmpstr** [comment]
> And it may also have a timeout?

The sync decoding case would not have a timeout. FWIW, the default decoding value is 'auto' and in Chromium that should be equivalent to 'sync' (iirc), so when we're talking about decoding=sync, that's just the default state

**Comment by mmocny** [comment]
Thanks @vmpstr!

---

Regarding a question @sebmarkbage asked up top, about how this relates to the `font-display: critical` -- I was also curious. Here's my understanding (and I'd love corrections if its off).

There are a few distinct uses for the phrase "render-blocking":

1. [Render-blocking resources](https://web.dev/render-blocking-resources/) (which [seems](https://html.spec.whatwg.org/#render-blocking-mechanism) already [specced](https://html.spec.whatwg.org/#update-the-rendering))
2. Anything explicitly preventing updates to the presentation of animation frames via some other form of throttling (as in image decode or SET)
3. Very broadly, just delays to updates for typical animation frames, i.e. just doing lots of work (long-tasks on main, expensive raster work off-main, etc)

My read of the `font-display: critical` proposal is it was just asking explicitly about (1).

My read is that SET needs some way to spec (2).

And (3) I know I use use to explain overly long paint-timings (i.e. FCP, LCP, INP, etc) which are specifically due to rendering issues.

---

There is already the concept of `rendering opportunity` which is typically how we explain frame throttling -- but HTML update-the-rendering is generally underspecified in terms of parallel rendering.  I have previously [found this doc](https://github.com/chrishtr/rendering/blob/master/rendering-event-loop.md#rendering-in-parallel) to be useful to explain some of it.

We have existing issues in terms of paint timing and element timing spec due to underspecified nature of parallel animation frame updates, which we would like to also be able to address.

**Comment by khushalsagar** [comment]
SET is going to use render blocking to explain how rendering is suppressed. For cross-document navigations, suppressing rendering of the new Document until is ready for transition will be controlled by "render-blocking".

For the same-document JS API, I was planning to expand the text [here](https://html.spec.whatwg.org/multipage/webappapis.html#update-the-rendering:~:text=A%20browsing%20context%20has%20no%20rendering%20opportunities%20if%20its%20active%20document%20is%20render%2Dblocked) to include [transition suppressing rendering](https://drafts.csswg.org/css-view-transitions-1/#transition-suppressing-rendering) as one of the reasons a document doesn't have rendering opportunities.

The spec language will need to ensure a transition suppresses rendering for all nested documents as well.

Also this was recently discussed in the issue [here](https://github.com/w3c/csswg-drafts/issues/7784) which has a summary of the discussion and the resolution.

**Comment by khushalsagar** [comment]
Closing this issue given the resolution on the csswg issue. Please feel free to continue the discussion there if there are more questions.

**Comment by yoavweiss** [comment]
> [!NOTE]
> 🤖 **This comment was written and posted by an AI agent** (Claude, running in the [Pi](https://github.com/badlogic/pi-mono) coding agent), acting on behalf of @yoavweiss while triaging this repository ahead of archiving it. The triage was derived from the current spec text and from open issues in `w3c/csswg-drafts`. **If it's wrong, please say so and it will be reopened** — no hard feelings, it's a bot.

**The titled question is resolved.** [`rendering suppression for view transitions`](https://drafts.csswg.org/css-view-transitions-2/#document-rendering-suppression-for-view-transitions) is now a defined per-`Document` concept, including the rule that while it is true, all pointer hit testing must target the document element. Follow-up editorial cleanup is w3c/csswg-drafts#10045.

⚠️ **However** — @sebmarkbage, the actual proposal in your OP (automatically delay playing the animation for up to ~100ms when entering content is still waiting on a font) was never adopted as a UA heuristic. It is now achievable explicitly:

- **same-document:** `await document.fonts.ready` (or a specific `FontFace.load()`) inside the `startViewTransition` update callback — the callback is allowed to be async precisely so you can gate on this;
- **cross-document:** [`<link rel=expect>`](https://html.spec.whatwg.org/multipage/links.html#link-type-expect) render-blocking on the new document (whatwg/html#9970).

If you still want the *automatic* behaviour, that's worth filing against csswg-drafts as a fresh issue. Closing this one, since the title's question is answered.

---

<sub>**Where this work lives now:** View Transitions is specified in [css-view-transitions-1](https://drafts.csswg.org/css-view-transitions-1/) and [css-view-transitions-2](https://drafts.csswg.org/css-view-transitions-2/), maintained in [w3c/csswg-drafts](https://github.com/w3c/csswg-drafts/issues?q=is%3Aissue+is%3Aopen+label%3Acss-view-transitions-1%2Ccss-view-transitions-2). Please file new issues there.</sub>
